web/src/lib/components/shared-components/combobox.svelte

```diff
@@ -11,81 +11,204 @@
 
 <script lang="ts">
   import { fly } from 'svelte/transition';
-
   import Icon from '$lib/components/elements/icon.svelte';
-  import { clickOutside } from '$lib/utils/click-outside';
   import { mdiMagnify, mdiUnfoldMoreHorizontal, mdiClose } from '@mdi/js';
-  import { createEventDispatcher } from 'svelte';
+  import { createEventDispatcher, tick } from 'svelte';
   import IconButton from '../elements/buttons/icon-button.svelte';
+  import type { FormEventHandler } from 'svelte/elements';
+  import { shortcuts } from '$lib/utils/shortcut';
+  import { onMount, onDestroy } from 'svelte';
 
-  export let id: string | undefined = undefined;
+  export let label: string;
+  export let hideLabel = false;
   export let options: ComboBoxOption[] = [];
   export let selectedOption: ComboBoxOption | undefined;
   export let placeholder = '';
 
+  /**
+   * Indicates whether or not the dropdown autocomplete list should be visible.
+   */
   let isOpen = false;
-  let inputFocused = false;
+  /**
+   * Keeps track of whether the combobox is actively being used.
+   */
+  let isActive = false;
   let searchQuery = selectedOption?.label || '';
+  let selectedIndex: number | undefined;
+  let comboboxRef: HTMLElement;
+  let optionRefs: HTMLElement[] = [];
+  const inputId = `combobox-${crypto.randomUUID()}`;
+  const listboxId = `listbox-${crypto.randomUUID()}`;
 
   $: filteredOptions = options.filter((option) => option.label.toLowerCase().includes(searchQuery.toLowerCase()));
 
+  onMount(() => {
+    window.addEventListener('click', onClick);
+  });
+
+  onDestroy(() => {
+    window.removeEventListener('click', onClick);
+  });
+
   const dispatch = createEventDispatcher<{
     select: ComboBoxOption | undefined;
     click: void;
   }>();
 
-  const handleClick = () => {
-    searchQuery = '';
+  const activate = () => {
+    isActive = true;
+    openDropdown();
+  };
+
+  const deactivate = () => {
+    searchQuery = selectedOption ? selectedOption.label : '';
+    isActive = false;
+    closeDropdown();
+  };
+
+  const openDropdown = () => {
     isOpen = true;
-    inputFocused = true;
-    dispatch('click');
   };
 
-  let handleOutClick = () => {
-    // In rare cases it's possible for the input to still have focus and
-    // outclick to fire.
-    if (!inputFocused) {
-      isOpen = false;
+  const closeDropdown = () => {
+    isOpen = false;
+    selectedIndex = undefined;
+  };
+
+  const moveFocus = (event: KeyboardEvent) => {
```

Comment:
What does this function do exactly? List elements aren't focusable because that's managed with `aria-activedescendant` and moving focus to the next/previous element should be default browser behavior.

Reply:
Found a solution for this - making the `preventDefault()` in the shortcuts action configurable with a flag.